src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -232,28 +244,34 @@ public async Task<bool> WaitFor100ContinueAsync(CancellationToken cancellationTo
                 return sendRequestContent;
             }
 
-            private void CheckForCompletion()
+            private void Reset()
             {
                 Debug.Assert(Monitor.IsEntered(SyncObject));
-                Debug.Assert(_requestCompletionState != StreamCompletionState.InProgress || _responseCompletionState != StreamCompletionState.InProgress,
-                    $"CheckForCompletion called but neither request nor response is completed");
-
-                if (_requestCompletionState == StreamCompletionState.InProgress || _responseCompletionState == StreamCompletionState.InProgress)
+                Debug.Assert(_requestCompletionState != StreamCompletionState.InProgress);
+                Debug.Assert(_responseCompletionState != StreamCompletionState.InProgress);
+                Debug.Assert(_requestCompletionState == StreamCompletionState.Failed || _responseCompletionState == StreamCompletionState.Failed,
+                    "Reset called but neither request nor response is failed");
+
+                if (NetEventSource.IsEnabled) Trace($"Stream reset. This is a test. Request={_requestCompletionState}, Response={_responseCompletionState}.");
+                // Don't send a RST_STREAM if we've already received one from the server.
+                if (_resetException == null)
```

Comment:
From a quick look at the source, I believe the answer is "yes", but we only ever set this while holding the same SyncObject lock, correct?  Just want to validate there's no chance we could see it's null, enter the if block, only to then have another thread set the exception.

Reply:
There's only once place this is set: during OnReset, and the SyncObject lock is held there.